pkg/api/v1/state.go

```diff
@@ -0,0 +1,232 @@
+	}
+	buf, err := ioutil.ReadAll(zlibReader)
+	if err != nil {
+		return nil, t, 0, fmt.Errorf("read: %v", err)
+	}
+	var g statepb.Grid
+	if err = proto.Unmarshal(buf, &g); err != nil {
+		return nil, t, 0, fmt.Errorf("parse: %v", err)
+	}
+	return &g, mod, gen, nil
+}
+
+// pathReader returns a reader for the specified path and last modified, generation metadata.
+func pathReader(ctx context.Context, client gcs.Client, path gcs.Path) (io.ReadCloser, time.Time, int64, error) {
+	r, attrs, err := client.Open(ctx, path)
+	if err != nil {
+		return nil, time.Time{}, 0, fmt.Errorf("client.Open(): %w", err)
+	}
+	if attrs == nil {
+		return r, time.Time{}, 0, nil
+	}
+	return r, attrs.LastModified, attrs.Generation, nil
+}
+
+// getDashboardTabInternal locate dashboard tab in config
+func getDashboardTabInternal(cfg *configpb.Configuration, dashboardKey string, tabKey string) *configpb.DashboardTab {
+	for _, dashboard := range cfg.Dashboards {
+		if config.Normalize(dashboard.Name) == dashboardKey {
+			for _, tab := range dashboard.DashboardTab {
+				if config.Normalize(tab.Name) == tabKey {
+					return tab
+				}
+			}
+			return nil
+		}
+	}
+	return nil
+}
+
+// getTabGrid fetch tab group name grid info (columns, rows, ..etc)
+func (s Server) getTabGrid(configPath *gcs.Path, cfg *configpb.Configuration, groupName string) (*statepb.Grid, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute) // todo(abdelrady): opt.timeout setup instead of fixed 10
+	defer cancel()
+	groupFinder := func(name string) (*configpb.TestGroup, gridReader, error) {
+		group := config.FindTestGroup(name, cfg)
+		if group == nil {
+			return nil, nil, nil
+		}
+		groupPath, err := configPath.ResolveReference(&url.URL{Path: path.Join(s.GridPathPrefix, name)})
+		if err != nil {
+			return group, nil, err
+		}
+		reader := func(ctx context.Context) (io.ReadCloser, time.Time, int64, error) {
+			return pathReader(ctx, s.Client, *groupPath)
+		}
+		return group, reader, nil
+	}
+
+	group, groupReader, err := groupFinder(groupName)
+	if err != nil || group == nil {
+		return nil, errors.New(fmt.Sprintf("not found: %q", groupName))
+	}
+	grid, _, _, err := readGrid(ctx, groupReader)
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("load %s: %v", groupName, err))
+	}
+
+	return grid, err
+}
+
+func (s Server) getDashboardTabGrid(r *http.Request, cfg *configpb.Configuration, dashboardKey string, tabKey string) (*statepb.Grid, string) {
+	tab := getDashboardTabInternal(cfg, dashboardKey, tabKey)
+	if tab == nil {
+		return nil, fmt.Sprintf("Dashboard {%q} or tab {%q} not found", dashboardKey, tabKey)
+	}
+	configPath, _, _ := s.configPath(r)
+	grid, err := s.getTabGrid(configPath, cfg, tab.TestGroupName)
+	if err != nil {
+		return nil, fmt.Sprintf("%v", err)
+	}
+	return grid, ""
+}
+
+// decodeRLE decode the run length encoded data
+//   [0, 3, 5, 4] -> [0, 0, 0, 5, 5, 5, 5]
+func decodeRLE(encodedData []int32) []int32 {
```

Comment:
We kiiiinda have this in https://github.com/GoogleCloudPlatform/testgrid/blob/c1f9367972d6fccb9a1d94112560acff469d3379/internal/result/results.go#L95, but not quite. Probably fine to keep this method as-is.

Reply:
Yes, Unfortunately the existing method doesn't provide required return type ([]int32) by the proto. So I will keep it for now.